tensorflow_encrypted/protocol/pond.py

```diff
@@ -756,16 +767,51 @@ def __init__(self, prot, variable0, variable1):
     def __repr__(self):
         return 'PondPrivateVariable(shape={})'.format(self.shape)
 
-def _encode(rationals, precision=BITPRECISION_FRACTIONAL):
-    """ [NumPy] Encode tensor of rational numbers into tensor of ring elements """
-    return (rationals * (2**precision)).astype(int).astype(object) % M
+def _encode(rationals, apply_scaling) -> BackingTensor:
+    """ Encode tensor of rational numbers into tensor of ring elements """
+
+    scaling_factor = 2**BITPRECISION_FRACTIONAL if apply_scaling else 1
+
+    if isinstance(rationals, np.ndarray):
+        return _encode_from_numpy(rationals, scaling_factor)
 
-def _decode(elements, precision=BITPRECISION_FRACTIONAL):
-    """ [NumPy] Decode tensor of ring elements into tensor of rational numbers """
+    if isinstance(rationals, tf.Tensor):
+        return _encode_from_tftensor(rationals, scaling_factor)
+
+    raise TypeError("Don't know how to encode {}".format(type(rationals)))
+
+def _decode(elements, apply_scaling: bool):
+    """ Decode tensor of ring elements into tensor of rational numbers """
+
+    scaling_factor = 2**BITPRECISION_FRACTIONAL if apply_scaling else 1
+
+    if isinstance(elements, np.ndarray):
+        return _decode_to_numpy(elements, scaling_factor)
+
+    if isinstance(elements, tf.Tensor):
+        return _decode_to_tftensor(elements, scaling_factor)
+
+    raise TypeError("Don't know how to decode {}".format(type(elements)))
+
+def _encode_from_numpy(rationals: np.ndarray, scaling_factor: int) -> BackingTensor:
+    encoded = (rationals * scaling_factor).astype(int).astype(object)
+    return BackingTensor.from_native(encoded)
+
+def _decode_to_numpy(elements, scaling_factor: int):
     map_negative_range = np.vectorize(lambda element: element if element <= M/2 else element - M)
-    return map_negative_range(elements).astype(float) / (2**precision)
+    return map_negative_range(elements).astype(float) / scaling_factor
+
+def _encode_from_tftensor(rationals: tf.Tensor, scaling_factor: int) -> BackingTensor:
+    encoded = tf.cast(rationals * scaling_factor, BackingTensor.int_type)
```

Comment:
How does this work?
Is there still the risk that we might overflow the `int_type`?

Reply:
not when the plaintext values are small enough (say integral part is < 2^16). the problem is in decoding since the current method for that uses large numbers